### PR TITLE
Fill missing real/virtual categories with 0

### PR DESCRIPTION
**Pull request by maurolepore**
@Souad-2DII,

This PR fills with `0` the values for missing categories (i.e. not real or virtual). This approach comes from a suggestion in previous conversations. But we still seem to have a problem: 

When the summary is calculated we sometimes end up with a `0/0` division. What should we do in such cases?

``` r
devtools::load_all()
#> ℹ Loading tacApp

# Values for missing categories are now filled with 0
prep <- prep_raw(full(), full()[1,])
#> Warning in min(data_portfolio$id): no non-missing arguments to min; returning
#> Inf
#> Warning in max(data_portfolio$id): no non-missing arguments to max; returning
#> -Inf
prep
#> # A tibble: 8 × 7
#>   technology    id category  type      start     end   value
#>   <chr>      <int> <chr>     <chr>     <dbl>   <dbl>   <dbl>
#> 1 renewables     1 before    total   0       0.00073 0.00073
#> 2 renewables     2 after     total   0.00073 0       0.00073
#> 3 renewables     3 add       real    0       0       0      
#> 4 renewables     4 remove    real    0       0       0      
#> 5 renewables     5 ramp up   real    0       0       0      
#> 6 renewables     6 ramp down real    0       0       0      
#> 7 renewables     7 buy       virtual 0       0       0      
#> 8 renewables     8 sell      virtual 0       0       0

# Real % is not a number. Is this acceptable?
summary <- summarize_change(prep)
summary
#> # A tibble: 1 × 5
#>   total_change real_change real_percent virtual_change virtual_percent
#>          <dbl>       <dbl>        <dbl>          <dbl>           <dbl>
#> 1            0           0          NaN              0             NaN

# Is the resulting plot valid?
plot_techs(prep)
```

<sup>Created on 2021-12-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

File: R/prep_raw.R [hunk 1]
```diff
@@ -35,3 +34,21 @@ prep_raw <- function(data, row = fake_row()) {
...
+        category %in% real_categories()    ~ "real",
+        category %in% virtual_categories() ~ "virtual",
+        TRUE                               ~ type
+      )
+    ) %>%
+    mutate(across(c(start, end, value), tidyr::replace_na, 0)) %>%
+    # FIXME: Do we need `id`?
+    select(-id) %>%
+    tibble::rowid_to_column(var = "id")
+}
```
Comment (maurolepore) [inline R/prep_raw.R:54]: Here is how this PR fills the values for the missing categories.

File: R/prepare_data_waterfall.R [hunk 2]
```diff
@@ -18,11 +18,12 @@ prepare_data_waterfall <- function(data,
...
+  # TODO: Remove dead code
   company_type_filter <- 1
 
   data_portfolio <- data %>%
     filter(.data$dual == 1) %>% # for now we only look at not dual plants as we don't know how to deal with dual - Angelika
-    # TODO: Ask Souad if this is dead code
+    # TODO: Remove dead code?
     #    filter(.data$CA100 == company_type_filter) %>%
     #    filter(.data$SBTI == company_type_filter) %>%
     #    filter(.data$comp_region == region_filter_2) %>%
```
Comment (maurolepore) [inline R/prepare_data_waterfall.R:29]: @Souad-2DII, while we're here -- can we delete the code that is commented out? There is more in the same file.

**Comment by Souad-2DII** [comment]
@maurolepore, yes sure go ahead and delete the code commented out.

**Comment by maurolepore** [comment]
@Souad-2DII, do you have a suggestion for what to do in the case shown in the first comment above? That is, when we prepare the raw data I can fill with `0` the values for missing categories. But then the summary table and plot are useless. What should we do?:

1. We are okay showing a useless table and plot.
2. We should instead throw an error.
3. We should do something else.

image

**Comment by Souad-2DII** [comment]
@maurolepore, this is the case when there is no data on the selected technology for the selected company (eg. the company does not produce electricity from coal only from renewables, so no data on coal).
In that case, it is better to show a message that says:
"There is no data on the selected technology", and not show neither the plot nor the tables.
What do you think?

**Comment by maurolepore** [comment]
I like your idea. Can you explain a bit more how you determine when there is no data for a techonlogy?

1. When the prepared dataset has no row for **any** of the real or virtual categories.
2. When the prepared dataset has no row for **some** of the real or virtual categories.
3. Something else.

**Comment by Souad-2DII** [comment]
When the real and virtual rows are empty (= 0) we display the "no data on the selected technology" message.
As long as there is a value for one of these categories we display the plot (even if the values are super low).

**Comment by maurolepore** [comment]
Closing in favor of https://github.com/2DegreesInvesting/tacApp/issues/3

**Pull request closed without merging**